Add automatic build script and Docker support

moisesPompilio opened the pull request:
### What was added:
**Automatic Build Script** (`auto_build.sh`):
- Introduced a script to automate module builds using `CXXFLAGS`.
- Supports optional cleaning modes (`FULL`, `CLEAN`, or selective cleaning) before building.
- Helps reduce manual steps and avoids repetitive commands.

**Docker Support:**
- Added a `Dockerfile` that allows running the fuzzer in an isolated environment without installing dependencies locally.
- Supports passing environment variables like `FUZZ`, `CXXFLAGS`, `FUZZ_RUNS`, and `FUZZ_INPUT`.
- Optionally mounts a volume to persist the generated corpus.

**Docker Compose Support:**
- Added docker-compose.yml for running multiple preconfigured fuzzing scenarios easily.
- Each service corresponds to a target, and the corpus is saved automatically in dedicated folders.
- Supports optional input and fuzzing run configuration via environment variables.

These improvements are now documented in the README.md, making it easier for developers to choose between manual builds, automated scripts, or containerized environments.

moisesPompilio committed:
build: add auto_build.sh for modular compilation with optional clean

This script automates the build process based on CXXFLAGS-defined modules.
It also offers an option to clean previous builds before compiling.

moisesPompilio committed:
build: add Docker support for fuzzing scenarios

Adds Dockerfile and docker-compose.yml to facilitate running fuzzing targets in isolated environments.
Supports environment variables FUZZ, CXXFLAGS, FUZZ_RUNS, and FUZZ_INPUT.
Corpus data is persisted via volume mounts to the host.

moisesPompilio commented:
Rebase completed

erickcestari commented on auto_build.sh:
```diff
@@ -0,0 +1,130 @@
+}
+
+# Define the list of modules
+modules="bitcoin_core rust_bitcoin rust_miniscript btcd embit lnd ldk nlightning"
```
It's missing the new module `clightning`

moisesPompilio replied:
Done

erickcestari commented on auto_build.sh:
```diff
@@ -0,0 +1,130 @@
+for module in $modules; do
+    module_name=$(echo "$module" | tr '[:lower:]' '[:upper:]')
+    if echo "$CXXFLAGS" | grep -q "$module_name"; then
+        case "$module" in
+            rust_bitcoin|rust_miniscript|ldk)
+                $module "rustup default nightly && make cargo && make"
+                ;;
+            *)
+                $module "make"
+                ;;
```
I think here we should simplify these rust projects to just use `make` to be like other modules. But we can do this in other PR since this would change Ci and Makefiles.

@brunoerg what do you think?

brunoerg replied:
We can leave it as-is. @i-am-yuvi is working on migrating from autotools to cmake and maybe it will be better.

erickcestari commented on docker-compose.yml:
```diff
@@ -0,0 +1,108 @@
+      dockerfile: Dockerfile
+    environment:
+      <<: *common-env
+      CXXFLAGS: "-DLDK -DLND -DNLIGHTNING"
```
```suggestion
      CXXFLAGS: "-DLDK -DLND -DNLIGHTNING -DCLIGHTNING"
```
Add CLN here

moisesPompilio replied:
Done

erickcestari commented on docker-compose.yml:
```diff
@@ -0,0 +1,108 @@
+      FUZZ: psbt_parse
+      ASAN_OPTIONS: detect_leaks=0
+    volumes:
+      - ./docker/psbt_parse:/app/data
```
It's missing the new target `deserialize_offer`

moisesPompilio replied:
Done

erickcestari commented on README.md:
````diff
@@ -43,98 +43,219 @@ Or install the complete boost library with
-git submodule update --init --recursive external/lightning
-cd modules/clightning
-make
-export CXXFLAGS="$CXXFLAGS -DCLIGHTNING"
-```
+    ```bash
+    cd modules/nlightning
+    make
+    export CXXFLAGS="$CXXFLAGS -DNLIGHTNING"
+    ```
 
-Once the modules are compiled, you can compile bitcoinfuzz and execute it:
-```bash
-make
-FUZZ=target_name ./bitcoinfuzz
-```
````
It's removing the C-lighting Manual Method

moisesPompilio replied:
Fixed

erickcestari commented on README.md:
````diff
@@ -43,98 +43,219 @@ Or install the complete boost library with
+    cd modules/mako
+    export MAKO_LIB_PATH="path/to/libmako.a"
+    make
+    ```
````
We removed `mako`

moisesPompilio replied:
Adjusted

moisesPompilio committed:
doc: add build instructions for auto_build.sh and Docker
Added documentation to README.md detailing the usage of auto_build.sh for automated module compilation and cleaning.
Included instructions for building and running the application using Docker and Docker Compose, with examples for environment variables and volume mounts.

erickcestari commented:
One thing that would be nice is to have a volume to save crash files on the user's machine. Right now, when the container stops due to a bitcoinfuzz crash (after finding a discrepancy), we lose the crash file.

erickcestari commented on auto_build.sh:
```diff
@@ -0,0 +1,134 @@
+clightning() {
+    execute_in_module "modules/clightning" "$1"
+}
+
```
The -DCUSTOM_MUTATOR_BOLT11 flag is missing. Which is basically building bitcoin core

moisesPompilio replied:
Done

erickcestari commented:
I had some trouble when trying to compile clightning. Probably the problem is the version of `glibc`. After bumping the version ubuntu 22.04 to -> 24.04 fixed the problem. Obs: I changed the python dependencies too, since it works different in 24.04

```
# Uses Ubuntu as base
FROM ubuntu:24.04

# Sets environment variables for installation
ENV DEBIAN_FRONTEND=noninteractive

# Installs basic tools
RUN apt-get update && apt-get install -y \
    wget curl git build-essential sudo \
    libc6-dev libgcc-11-dev libasan6 \
    cmake \
    && rm -rf /var/lib/apt/lists/*

# Installs Rust
RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
ENV PATH="/root/.cargo/bin:${PATH}"
RUN rustup install stable && rustup install nightly

# Installs Go
RUN wget https://golang.org/dl/go1.22.2.linux-amd64.tar.gz && \
    tar -C /usr/local -xzf go1.22.2.linux-amd64.tar.gz && \
    rm go1.22.2.linux-amd64.tar.gz
ENV PATH="/usr/local/go/bin:${PATH}"

# Installs .NET SDK 8.0
RUN wget https://packages.microsoft.com/config/ubuntu/22.04/packages-microsoft-prod.deb -O packages-microsoft-prod.deb && \
    dpkg -i packages-microsoft-prod.deb && \
    rm packages-microsoft-prod.deb && \
    apt-get update && apt-get install -y dotnet-sdk-8.0 && \
    rm -rf /var/lib/apt/lists/*

# Installs Python 3.11 and pip
RUN apt-get update && apt-get install -y python3 python3-pip python3-venv && \
    rm -rf /var/lib/apt/lists/*

# Installs LLVM and Clang 18
RUN apt-get update && apt-get install -y \
    lsb-release wget software-properties-common gnupg && \
    wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add - && \
    add-apt-repository "deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-18 main" && \
    apt-get update && apt-get install -y \
    clang-18 clang++-18 && \
    ln -sfT /usr/bin/clang++-18 /usr/bin/clang++ && \
    ln -sfT /usr/bin/clang-18 /usr/bin/clang && \
    rm -rf /var/lib/apt/lists/*

# Installs Boost
RUN apt-get update && apt-get install -y libboost-all-dev && \
    rm -rf /var/lib/apt/lists/*

# Configures environment variables
ENV CC=/usr/bin/clang
ENV CXX=/usr/bin/clang++
ENV BOOST_LIB_DIR=/usr/lib/x86_64-linux-gnu/
ENV LDFLAGS="-lsodium"

# Working directory
WORKDIR /app

# Copies the repository
COPY . .

# Creates a virtual environment
RUN python3 -m venv /venv
ENV PATH="/venv/bin:$PATH"

# Installs dependencies for embit
RUN pip install -r modules/embit/requirements.txt

# Installs dependencies for C-lightning
RUN python3 -m pip install --upgrade pip && \
    python3 -m pip install mako
RUN git submodule update --init --recursive external/lightning
RUN apt-get update && apt-get install -y \
    libsqlite3-dev \
    libsodium-dev \
    jq \
    && rm -rf /var/lib/apt/lists/*

# Clean cache bitcoin core
RUN rm -rf /app/modules/bitcoin/univalue/build && \
    mkdir -p /app/modules/bitcoin/univalue/build && \
    cd /app/modules/bitcoin/univalue && \
    CXXFLAGS="-fsanitize=address" cmake -B build && \
    cmake --build build && \
    cd build && \
    make univalue

# Copies and gives permission to auto_build.sh
COPY auto_build.sh .
RUN chmod +x auto_build.sh

# CMD requires FUZZ, but FUZZ_RUNS and FUZZ_INPUT are optional
CMD ["bash", "-c", "mkdir -p /app/data/crash && \
    ./auto_build.sh && \
    if [ -z \"$FUZZ\" ]; then \
    echo \"Error: FUZZ not defined\"; \
    exit 1; \
    elif [ -n \"$FUZZ_INPUT\" ]; then \
    ./bitcoinfuzz -artifact_prefix=/app/data/crash/ \"$FUZZ_INPUT\"; \
    elif [ -n \"$FUZZ_RUNS\" ]; then \
    ./bitcoinfuzz -runs=$FUZZ_RUNS -artifact_prefix=/app/data/crash/ /app/data; \
    else \
    ./bitcoinfuzz -artifact_prefix=/app/data/crash/ /app/data; \
    fi"]
```

erickcestari reviewed (approved):
ACK 40575f8

yuvicc commented on docker-compose.yml:
```diff
@@ -0,0 +1,119 @@
+      dockerfile: Dockerfile
+    environment:
+      <<: *common-env
+      CXXFLAGS: "-DEMBIT -DRUST_BITCOIN -DBTCD"
```
We now also support bitcoin-core here,

`-DBITCOIN_CORE`

moisesPompilio replied:
Done!

yuvicc reviewed (commented):
Also while reviewing this pr, I found that we don't execute `psbt_parse` target for embit in workflow file. You can add that to the next commit or a separate pr for that.

see [here](https://github.com/bitcoinfuzz/bitcoinfuzz/blob/e6e6412a6537577a70e78b94d70684da670048a0/.github/workflows/workflow.yml#L195)

yuvicc commented:
Please rebase as we have added some trophies in readme.

yuvicc commented on README.md:
```diff
@@ -43,98 +43,219 @@ Or install the complete boost library with
+     ├── miniscript_parse/
+     ├── script_asm/
+     ├── deserialize_invoice/
+     ├── address_parse/
```
nit: `addrv2` as well!

moisesPompilio replied:
done!

yuvicc reviewed (changes requested):
Please rebase as we have added some trophies in readme.

brunoerg reviewed (approved):
light code review ACK 002122c2cadf1a664b7b78a698b7a9b1e486b26b

any improvement could be done in a follow-up PR.
